src/passes/TraceCalls.cpp

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Instruments the build with code to intercept selected function calls.
+// This can be e.g. used to trace allocations (malloc, free, calloc, realloc)
+// and build tools for memory usage analysis.
+// The pass supports SIMD but multi-value feature is not supported yet.
+//
+// Instrumenting void free(void*):
+//  Before:
+//   (call $free (i32.const 64))
+//
+//  After:
+//   (local $1 i32)
+//   (block
+//     (call $free
+//       (local.tee $1
+//         (i32.const 64)
+//       )
+//     )
+//     (call $trace_free
+//       (local.get $1)
+//     )
+//   )
+//
+// Instrumenting void* malloc(int32_t) with a user-defined tracer (trace_allocation):
+//  Before:
+//   (call $malloc
+//     (local.const 32))
+//
+//  After:
+//   (local $0 i32)
+//   (local $1 i32)
+//   (block (result i32)
+//     (call $trace_allocation
+//       (local.get $0)
+//       (local.tee $1
+//         (call $malloc
+//           (local.tee $0 (i32.const 2))
+//         )
+//       )
+//     )
+//   )
+//
+
+#include <set>
+
+#include "asmjs/shared-constants.h"
+#include "ir/import-utils.h"
+#include "pass.h"
+#include "support/string.h"
+#include "wasm-builder.h"
+
+namespace wasm {
+
+struct TracedFunction {
+  Name originName;
+  Name tracerName;
+  Type params;
+  Type results;
+};
+bool operator<(const TracedFunction& lhs, const TracedFunction& rhs) {
+  return lhs.originName < rhs.originName;
+}
+
+static Type strToType(const std::string& strType) {
+  if (strType.compare("i32") == 0)
+    return Type::i32;
+  else if (strType.compare("i64") == 0)
+    return Type::i64;
+  else if (strType.compare("f32") == 0)
+    return Type::f32;
+  else if (strType.compare("f64") == 0)
+    return Type::f64;
+  else if (strType.compare("v128") == 0)
+    return Type::v128;
+  else
+    Fatal() << "Failed to parse type '" << strType << "'";
+}
+
+struct AddTraceWrappers : public WalkerPass<PostWalker<AddTraceWrappers>> {
+  AddTraceWrappers(std::set<TracedFunction> tracedFunctions)
+    : tracedFunctions(std::move(tracedFunctions)) {}
+  void visitCall(Call* curr) {
+    auto* target = getModule()->getFunction(curr->target);
+
+    auto function_it = std::find_if(tracedFunctions.begin(),
```

Comment:
`function_it` etc. should be `functionIt` in terms of naming conventions. However, in this case it can be just `iter` which is what we tend to use for a quick iterator.